### DISCUSSION TITLE
Нужна помощь! Ошибка при проверке подписи: The hash value is not correct.

### DISCUSSION THREAD

**Discussion by DmitryKrakhmalev** [CryptoPro/libcore#43] [Q&A] [answered]
Добрый день. Помогите разобраться в чем может заключаться проблема валидации открепленной подписи определенных файлов. Практически все остальные файлы валидируются без ошибок.
[Файл и подпись.zip](https://github.com/CryptoPro/libcore/files/13388245/default.zip)

Валидирую через SignedCms -CheckSignature(true), в момент валидации возникает ошибка.

<details>
  <summary>Описание ошибки: The hash value is not correct.</summary>
  
```cs
System.Security.Cryptography.CryptographicException
  HResult=0x80131501
  Message=The hash value is not correct.
  Source=LibCore.Security.Cryptography.Pkcs
  StackTrace:
   at LibCore.Security.Cryptography.Pkcs.Detours.SignerInfoDetour.PrepareDigest(Boolean compatMode, SignerInfo instance)
   at LibCore.Security.Cryptography.Pkcs.Detours.SignerInfoDetour.VerifySignature(GostCmsSignature signatureProcessor, X509Certificate2 certificate, Boolean compatMode, SignerInfo instance)
   at LibCore.Security.Cryptography.Pkcs.Detours.SignerInfoDetour.Prefix(X509Certificate2Collection extraStore, X509Certificate2 certificate, Boolean verifySignatureOnly, SignerInfo __instance)
   at System.Security.Cryptography.Pkcs.SignerInfo.CheckSignature(X509Certificate2Collection extraStore, Boolean verifySignatureOnly)
   at System.Security.Cryptography.Pkcs.SignedCms.CheckSignatures(SignerInfoCollection signers, X509Certificate2Collection extraStore, Boolean verifySignatureOnly)
   at System.Security.Cryptography.Pkcs.SignedCms.CheckSignature(X509Certificate2Collection extraStore, Boolean verifySignatureOnly)
   at System.Security.Cryptography.Pkcs.SignedCms.CheckSignature(Boolean verifySignatureOnly)
   at CryptoTools.Controllers.CryptoController.<VerifySign>d__4.MoveNext() in D:\Projects\SaInSpUtils.DevExpress\CryptoTools\Controllers\CryptoController.cs:line 117

  This exception was originally thrown at this call stack:
    LibCore.Security.Cryptography.Pkcs.Detours.SignerInfoDetour.PrepareDigest(bool, System.Security.Cryptography.Pkcs.SignerInfo)
    LibCore.Security.Cryptography.Pkcs.Detours.SignerInfoDetour.VerifySignature(LibCore.Security.Cryptography.GostCmsSignature, System.Security.Cryptography.X509Certificates.X509Certificate2, bool, System.Security.Cryptography.Pkcs.SignerInfo)
    LibCore.Security.Cryptography.Pkcs.Detours.SignerInfoDetour.Prefix(System.Security.Cryptography.X509Certificates.X509Certificate2Collection, System.Security.Cryptography.X509Certificates.X509Certificate2, bool, System.Security.Cryptography.Pkcs.SignerInfo)
    System.Security.Cryptography.Pkcs.SignerInfo.CheckSignature(System.Security.Cryptography.X509Certificates.X509Certificate2Collection, bool)
    System.Security.Cryptography.Pkcs.SignedCms.CheckSignatures(System.Security.Cryptography.Pkcs.SignerInfoCollection, System.Security.Cryptography.X509Certificates.X509Certificate2Collection, bool)
    System.Security.Cryptography.Pkcs.SignedCms.CheckSignature(System.Security.Cryptography.X509Certificates.X509Certificate2Collection, bool)
    System.Security.Cryptography.Pkcs.SignedCms.CheckSignature(bool)
    CryptoTools.Controllers.CryptoController.VerifySign(System.Guid, System.Guid) in CryptoController.cs
```
  
</details>

<details>
   <summary>Технические подробности</summary>

Используется LibCore версии **2023.8.9.1**.
Крипто Про CSP с лицензией версия **5.0.12922**
![image](https://github.com/CryptoPro/libcore/assets/11774163/9b8e3520-ce26-43c3-b375-c149b719f401)

</details>

Подпись действительна, валидируется другими средствами. Такая ошибка возникла впервые, остальные подписи валидировались. Изначально подумал что проблема в корневых сертификатах, установил корневые сертификаты.
![image](https://github.com/CryptoPro/libcore/assets/11774163/62813553-849c-41e2-b1e0-db0c0f2ad2e1)

Попробовал свалидировать подпись средствами cryptcp - все проверяется успешно.
![image](https://github.com/CryptoPro/libcore/assets/11774163/9ce4fb3d-d296-4931-b979-afcb758f779a)

Подскажите пожалуйста в чем проблема?

**Comment by Bykiev** [comment]
Здравствуйте, а что за странное условие с ASCII? Он идет по этой ветке или в условие else? Ошибка проверки подписи может быть связана с кодировкой

**Reply by DmitryKrakhmalev** [reply]
Это чтобы избавиться от _-----BEGIN CMS----- и -----END CMS-----_. Если этого не делать то при декодировании **verifyCms.Decode(signature)** возникает ошибка **"ASN1 corrupted data."**
Идет по этой ветке, декодирование успешно проходит, видно что в verifyCms имеются подписанты, что все хорошо. По тому же сценарию проходят проверку и другие файлы. Тоже было предположение что возможно проблема с кодировкой.
![image](https://github.com/CryptoPro/libcore/assets/11774163/5bdb19a5-1654-4326-84a3-acdb2c2e41e0)
![image](https://github.com/CryptoPro/libcore/assets/11774163/0e1d08e1-282f-459c-916f-1d4c3117b335)

**Reply by Bykiev** [reply]
Какая у вас версия LibCore, CSP, ОС?

**Reply by DmitryKrakhmalev** [reply]
Используется LibCore версии 2023.8.9.1.
Крипто Про CSP с лицензией версия 5.0.12922
ОС - тестирую на Windows LibCore  под Windows а крутится все на linux, там аналогичная проблема

**Comment by Fasjeit** [comment, marked as answer]
У вас **присоединённая** подпись (содержимое включается в саму подпись).

Подпись из вложения успешно проверяется так, явно указывать исходные подписываемые данные в `ContentInfo` не нужно:
```csharp
var signedCms = new SignedCms();
// DetachedData - base64 из файла без BEGIN/END и переносов строк
signedCms.Decode(Convert.FromBase64String(DetachedData));
signedCms.CheckSignature(false);
```

**Reply by DmitryKrakhmalev** [reply]
Спасибо большое! Работает. Попробую добавить в свой код определение типа открепленная или присоединенная подпись и на основании этого отрабатывать по разному
